### DISCUSSION TITLE
i have problem when using family and object with FamilyAsyncNotifier

### DISCUSSION THREAD

**Discussion by mustafa-707** [rrousselGit/riverpod#3388] [Q&A]
thats my code :

#on provider file 
```
final newsStateProvider =
    AsyncNotifierProvider.family<NewsStateController, NewsState, ArticlesInfo>(
        () {
  return NewsStateController();
});

class NewsStateController extends FamilyAsyncNotifier<NewsState, ArticlesInfo> {
  int currentPageNumber = 0;

  @override
  Future<NewsState> build(ArticlesInfo arg) async {
    late NewsContent future;
    if (arg.tagName != null) {
      future = await NewsServices.getListOfTagsArticles(
        arg.countryIso2,
        arg.pageNumber,
        arg.tagName!,
      );
    } else {
      future = await NewsServices.getListOfArticles(
        arg.countryIso2,
        arg.pageNumber,
        arg.withVideos,
        arg.sectionsId ?? [],
      );
    }
    return NewsState(
      mainArticle: future.sectionContent.json?.mainArticle,
      articles: future.sectionContent.json?.list ?? [],
      ads: future.sectionpageAds,
    );
  }
}

```
#on Widget
```
  @override
  Widget build(BuildContext context) {
    final info = ArticlesInfo(
       withVideos: widget.withVideos,
       countryIso2: widget.location,
       sectionsId: widget.listOfIds,
     );

     final provider = newsStateProvider(info);
    final fetchListOfArticles = ref.watch(newsStateProvider);
    return SingleChildScrollView(
          physics: const ClampingScrollPhysics(), //disable bouncing
          child: switch (fetchListOfArticles) {
            AsyncData(:final value) => value.articles.isEmpty
                ? const NoNewsListWidget()
                : Column(
                    children: <Widget>[
                      if (value.mainArticle != null)
                        MainNodeWidget(
                          data: value.mainArticle!,
                        ),
                      Visibility(
                        visible: value.ads != null,
                        child: old_provider.Consumer<MainProvider>(
                          builder: (context, main, _) {
                            if (main.isPremium) {
                              return const SizedBox();
                            } else {
                              return AwMpuAds(
                                hmsAdUnitId: value.ads!.adunit2.adUnitIdHms,
                                place: "news_content",
                                key: const PageStorageKey("news_content"),
                                adUnitId: value.ads!.adunit2.adUnitId,
                                adSize: value.ads!.adunit2.adSize,
                                adType: value.ads!.adunit2.adType,
                              );
                            }
                          },
                        ),
                      ),
                      NodesList(
                        nodesList: value.articles,
                      ),
                      isLoading
                          ? const AWSkeletonLoaderArticleListingPage()
                          : const SizedBox.shrink(),
                    ],
                  ),
            AsyncError() => SizedBox(
                width: MediaQuery.of(context).size.width,
                height: MediaQuery.of(context).size.height - 340,
                child: AwErrorWidget(
                  onTap: () {
                    // ref.invalidate(provider);
                  },
                ),
              ),
            _ => const Center(child: AWSkeletonLoaderArticleListingPage()),
          },
        );
        ....
```

the expected behavior to build initial value 

whats actually happens is trigger infinite loop of building first inital values

**Comment by Ahmadajami** [comment]
Same for me do you fix it you

**Reply by mustafa-707** [reply]
Not yet, but it get more annoying.

**Comment by gdurandrexel** [comment]
You must not create a new instance of ArticlesInfo in your build method: build is called on every change of state, in this case it builds at least twice, once when the fetchListOfArticles is AsyncLoading and another time when it is AsyncData. In the second build call you create a new instance of ArticlesInfo that you pass as parameter to the family provider. Since these 2 instances are different (and since you probably did not override == and hashCode), a new provider is created, which triggers a new loading cycle.
So basically, each time you build, you create a new provider that triggers a new build, etc.

So either override == and hashCode in ArticlesInfo or create only one instance in your widget from withVideos, location, and listOfIds (probably more what you want).
